clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java

```diff
@@ -225,11 +230,20 @@ public double compressionRate() {
     }
 
     public boolean isFull() {
-        return recordsBuilder.isFull();
+        return !isWritable || recordsBuilder.isFull();
```

Comment:
Hmm. You may be right. The reason I introduced `isWritable` is because we can't close a batch when it is full. we can only close it at the point of sending so that we set the right sequence number. So I introduced `isWritable` to denote the state where it can no longer take appends, but is not closed. 

Your point is valid: once it is considered full for a particular append, no future appends should go to it since there will be another batch at the tail which should get the new appends.